### PR TITLE
feat(snql): SnQL Grammar and Visitor, Part 2

### PR DESCRIPTION
**Pull request by ayirr7**
Updates:

- Now supports conditions (WHERE and HAVING clauses) along with previous clauses
- An added test case shows slightly more complicated combination of conditions, with both AND / OR
- Added more type hints, comments

Next steps:
- Any revisions
- Adding parentheses (and priorities) to grammar, visitor

### CODE REVIEW

File: snuba/query/snql/__init__.py [hunk 1]
```diff
@@ -29,19 +35,23 @@
     visit_parameters_list,
     visit_quoted_literal,
 )
-from typing import Sequence
-
 
 snql_grammar = Grammar(
     fr"""
-    query_exp             = match_clause where_clause? collect_clause? group_by_clause? having_clause? order_by_clause?
+    query_exp             = match_clause where_clause collect_clause group_by_clause? having_clause? order_by_clause?
```
Comment (lynnagara) [inline snuba/query/snql/__init__.py:41]: How come the where and collect clauses are now mandatory? Wouldn't a query without these still be valid?

File: snuba/query/snql/__init__.py [hunk 2]
```diff
@@ -29,19 +35,23 @@
...
     match_clause          = space* "MATCH" space* open_paren clause close_paren space*
-    where_clause          = space* "WHERE" clause space*
+    where_clause          = space* "WHERE" or_expression space*
     collect_clause        = space* "COLLECT" collect_list space*
     group_by_clause       = space* "BY" group_list space*
+    having_clause         = space* "HAVING" or_expression space*
     order_by_clause       = space* "ORDER BY" order_list space*
 
+    condition             = low_pri_arithmetic condition_op (column_name / numeric_literal)
+    condition_op          = "=" / "!=" / ">" / ">=" / "<" / "<="
```
Comment (lynnagara) [inline snuba/query/snql/__init__.py:51]: Should "IN" be on this list as well?

File: snuba/query/snql/__init__.py [hunk 3]
```diff
@@ -29,19 +35,23 @@
...
 
     match_clause          = space* "MATCH" space* open_paren clause close_paren space*
-    where_clause          = space* "WHERE" clause space*
+    where_clause          = space* "WHERE" or_expression space*
     collect_clause        = space* "COLLECT" collect_list space*
     group_by_clause       = space* "BY" group_list space*
+    having_clause         = space* "HAVING" or_expression space*
     order_by_clause       = space* "ORDER BY" order_list space*
 
+    condition             = low_pri_arithmetic condition_op (column_name / numeric_literal)
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:50]: I think you also need string literals on the right hand side of this expression.
Also thinking over this, I think we will need to support functions on the right hand side as well. 
Think about `timestamp > toDateTime('2020-01-01')`
I don't like making the language silently convert the string type in the datetime. We can introduce that conversion later on and limit the allowed functions on the right hand side on after parsing.

A good idea here would be to ask for opinions on the SnQL discussion channel. You could ask people there what they think, if they prefer `timestamp > '2020-01-01'` like we do today or if they prefer to provide the function call explicitly. In the meantime you can accept functions.

File: snuba/query/snql/__init__.py [hunk 4]
```diff
@@ -88,13 +97,15 @@ class SnQLVisitor(NodeVisitor):
...
+        _, where, collect, groupby, having, orderby = visited_children
         # check for empty clauses
         if isinstance(groupby, Node):
             groupby = None
         if isinstance(orderby, Node):
             orderby = None
-        return Query({}, None, collect, None, None, None, groupby, None, orderby)
+        if isinstance(having, Node):
+            having = None
+        return Query({}, None, collect, None, where, None, groupby, having, orderby)
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:108]: please assign these arguments by key, so what is null would be more clear.
```
return Query(
   body={},
   data_source=....
   ...
}

File: snuba/query/snql/__init__.py [hunk 5]
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
     ) -> Literal:
         return visit_quoted_literal(node, visited_children)
 
+    def visit_where_clause(
+        self, node: Node, visited_children: Tuple[Any, Any, Sequence[Expression], Any]
+    ) -> Sequence[Expression]:
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:167]: I think this should be returning a single Expression that represent the whole boolean conditions tree not a sequence of expressions.

File: snuba/query/snql/__init__.py [hunk 6]
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
...
+        _, _, conditions, _ = visited_children
+        return conditions
+
+    def visit_having_clause(
+        self, node: Node, visited_children: Tuple[Any, Any, Sequence[Expression], Any]
+    ) -> Sequence[Expression]:
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:173]: Same as above:
I think this should be returning a single Expression that represent the whole boolean conditions tree not a sequence of expressions.

File: snuba/query/snql/__init__.py [hunk 7]
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
...
+        return conditions
+
+    def visit_and_expression(
+        self, node: Node, visited_children: Tuple[Any, Expression, Any, Expression]
+    ) -> Expression:
+        _, left_condition, _, and_condition = visited_children
+        # in the case of one Condition
+        # and_condition will be an empty Node
+        if not isinstance(and_condition, Node):
+            _, right_condition = and_condition
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:184]: Isn't `and_condition` a sequence of multiple `["AND", Expression]` ? So shouldn't you iterate over right_condition ?
If not, I am not sure I get who is combining the the multiple conditions in AND.

File: snuba/query/snql/__init__.py [hunk 8]
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
...
+
+    def visit_or_expression(
+        self, node: Node, visited_children: Tuple[Any, Expression, Any, Expression]
+    ) -> Expression:
+        _, left_condition, _, or_condition = visited_children
+        # in the case of one Condition
+        # or_condition will be an empty Node
+        if not isinstance(or_condition, Node):
+            _, right_condition = or_condition
+            return combine_or_conditions([left_condition, right_condition])
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:196]: same as above

File: snuba/query/snql/__init__.py [hunk 9]
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
...
+        return left_condition
+
+    def visit_condition_comma(
+        self, node: Node, visited_children: Tuple[Expression, Any, Any, Any]
+    ) -> Expression:
+        bin_condition, _, _, _ = visited_children
+        return bin_condition
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:203]: What is this for ?

File: tests/query/snql/test_query.py [hunk 10]
```diff
@@ -31,16 +36,84 @@
...
+                                Column(None, None, "last_seen"),
+                                Literal(None, 2),
+                            ),
+                        ]
+                    ),
+                ],
+            ),
         ),
-        id="Simple COLLECT, GROUPBY, ORDERBY clause example",
+        id="Example 2",
```
Comment (fpacifici) [inline tests/query/snql/test_query.py:116]: I would suggest to add a few simpler tests that test some specific conditions:
1) multiple conditions in AND (more than 2). The other clauses can be trivial
2) multiple conditions in OR (more than 2). The other clauses can be trivial
3) Conditions with different types of literals (int, columns, strings)
4) Combination of AND and OR at multiple levels.

This approach will allow you to cover more cases without having a massive query each time to express with the AST.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Seems a good step in the right direction, please see the comments inline.

File: snuba/query/snql/__init__.py (hunk 9 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py:203]: Removed.

**Comment by ayirr7** [comment]
Updates:

- Modified the logic for dealing with conditions in Visitor
- Added function_call to right hand side of a condition in grammar
- Added test cases for various conditions/combinations of conditions
- Improved some type hints

Next Steps:

- PR for parentheses

File: snuba/query/snql/__init__.py (hunk 1 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py:40]: I think that makes sense - I gave it some thought and compared with Cypher query language as well - it seems to me that a query would still be valid without a WHERE or COLLECT?

File: snuba/query/snql/__init__.py (hunk 2 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py:50]: Currently, no, I'm handling the other conditional operators first. The next step is to deal with parenthetical expressions (those that indicate priority), and if time permits, I would deal with the other operators like "IN".

File: snuba/query/snql/__init__.py [hunk 11]
```diff
@@ -29,37 +34,46 @@
...
 
     match_clause          = space* "MATCH" space* open_paren clause close_paren space*
-    where_clause          = space* "WHERE" clause space*
+    where_clause          = space* "WHERE" or_expression space*
     collect_clause        = space* "COLLECT" collect_list space*
     group_by_clause       = space* "BY" group_list space*
+    having_clause         = space* "HAVING" or_expression space*
     order_by_clause       = space* "ORDER BY" order_list space*
 
+    condition             = low_pri_arithmetic condition_op (function_call / column_name / numeric_literal) space*
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:49]: I think you are missing the string literal.

File: snuba/query/snql/__init__.py [hunk 12]
```diff
@@ -29,37 +34,46 @@
...
+    condition_op          = "=" / "!=" / ">" / ">=" / "<" / "<="
+    interm_condition      = open_paren or_expression close_paren
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:51]: let's give this a better name. Like `parenthesized_condition` ?

File: snuba/query/snql/__init__.py [hunk 13]
```diff
@@ -29,37 +34,46 @@
...
-
     order_list            = order_columns* low_pri_arithmetic ("ASC"/"DESC")
     order_columns         = low_pri_arithmetic ("ASC"/"DESC") space* comma space*
 
-    having_clause         = space* "HAVING" clause space*
     clause                = space* ~r"[-=><\w]+" space*
 
     low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
-    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    high_pri_arithmetic   = (space* arithmetic_term space* (high_pri_tuple)*)
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:70]: why did you add parenthes here ?

File: tests/query/snql/test_query.py [hunk 14]
```diff
@@ -31,16 +32,254 @@
...
             groupby=[
                 Column(None, None, "d"),
                 FunctionCall(None, "plus", (Literal(None, 2), Literal(None, 7),),),
             ],
-            order_by=[
-                OrderBy(OrderByDirection.DESC, Column(None, None, "f")),
-                OrderBy(OrderByDirection.ASC, Column(None, None, "m")),
+            order_by=[OrderBy(OrderByDirection.DESC, Column(None, None, "f"))],
+        ),
+        id="Example 1",
```
Comment (fpacifici) [inline tests/query/snql/test_query.py:44]: Please provide a descriptive name to the test. It should convey what this test is for.

**Review by fpacifici — APPROVED** [review]
Please address the comments before merging.

File: tests/query/snql/test_query.py [hunk 15]
```diff
@@ -2,14 +2,15 @@
 
 from snuba import state
 from snuba.datasets.factory import get_dataset
+from snuba.query.conditions import binary_condition
 from snuba.query.expressions import Column, FunctionCall, Literal
 from snuba.query.logical import OrderBy, OrderByDirection, Query, SelectedExpression
 from snuba.query.snql import parse_snql_query
 
 
 test_cases = [
```
Comment (evanh) [inline tests/query/snql/test_query.py:11]: I think you should also start adding some negative tests, i.e. queries that shouldn't work.

For example, I'm curious what happens if I omit spaces from my query. Based on the regex, spaces seem to be totally optional.

Is `MATCH(blah)WHEREa<3COLLECT4-5` a valid query?

Comment (ayirr7) [reply tests/query/snql/test_query.py:11]: Hi, I've started adding some test cases that serve as invalid queries that hopefully begin to illustrate what sorts of queries cannot be parsed and why. Would appreciate any further suggestions to add more cases!

File: tests/query/snql/test_query.py (hunk 14 above)
Comment (ayirr7) [reply tests/query/snql/test_query.py]: Added.

File: snuba/query/snql/__init__.py (hunk 13 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: Removed.

File: snuba/query/snql/__init__.py (hunk 12 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: Will modify names in P3 (where the parenthesized conditions will be).

File: snuba/query/snql/__init__.py (hunk 11 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: Added.

File: snuba/query/snql/__init__.py [hunk 16]
```diff
@@ -29,37 +34,45 @@
...
     clause                = space* ~r"[-=><\w]+" space*
 
     low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
     high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
     low_pri_tuple         = low_pri_op high_pri_arithmetic
     high_pri_tuple        = high_pri_op arithmetic_term
-    arithmetic_term       = (space*) (function_call / numeric_literal / column_name) (space*)
+
+    arithmetic_term       = space* (function_call / numeric_literal / column_name / interm_arithm) space*
+    interm_arithm         = open_paren low_pri_arithmetic close_paren
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:74]: please give to this one a descriptive name.

File: snuba/query/snql/__init__.py [hunk 17]
```diff
@@ -139,6 +192,7 @@ def visit_high_pri_arithmetic(
         node: Node,
         visited_children: Tuple[Any, Expression, Any, HighPriArithmetic],
     ) -> Expression:
+
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:195]: please revert this change.

File: tests/query/snql/test_invalid_queries.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+test_cases = [
+    # below are all cases that are not parsed completely
+    # i.e. the entire string is not consumed
+    pytest.param(
+        "MATCH(blah)WHEREa<3COLLECT4-5,3*g(c),cBYd,2+7ORDERBYfDESC",
+        IncompleteParseError,
+        id="ORDER BY is two words",
+    ),
+    pytest.param(
+        "MATCH(blah)WHEREa<3COLLECT4-5,3*g(c),cBYd,2+7ORDER BYfDESC",
```
Comment (fpacifici) [inline tests/query/snql/test_invalid_queries.py:16]: mind adding the spaces between match, where and collect, if you are testing the ORDER BY, the rest of the query can still be valid.

File: tests/query/snql/test_invalid_queries.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+        IncompleteParseError,
+        id="Expression before ASC / DESC needs to be separated from ASC / DESC keyword by space",
+    ),
+    pytest.param(
+        "MATCH(blah)WHEREa<3COLLECT4-5,3*g(c),cBYd,,2+7 ORDER BY f DESC",
```
Comment (fpacifici) [inline tests/query/snql/test_invalid_queries.py:21]: same comment.
